File: lib/apps/triplet_loader2.dart
import 'dart:io';
import 'dart:math';
import 'dart:typed_data';
import 'package:image/image.dart' as img;

class TripletLoader {
  final String rootPath;
  final int imageSize;

  // 🔥 FIX 1: Change Map type to store the processed pixels, not the File pointers
  final Map<String, List<Float32List>> _identityMap = {};
  final Random _random = Random();

  TripletLoader(this.rootPath, this.imageSize, int numOfFiles) {
    _scanDataset(numOfFiles);
  }

  void _scanDataset(int numOfFiles) {
    int scannedFiles = 0;
    final rootDir = Directory(rootPath);
    if (!rootDir.existsSync()) return;

    print("Caching images to RAM as Float32List...");

    for (var entity in rootDir.listSync()) {
      if (entity is Directory) {
        final personName = entity.path.split(Platform.pathSeparator).last;

        final List<Float32List> processedImages = [];

        final files = entity.listSync().whereType<File>().where(
          (f) => f.path.endsWith('.jpg') || f.path.endsWith('.png'),
        );

        for (var file in files) {
          // 🔥 FIX 2: Process the file into pixels immediately
          processedImages.add(_processImage(file));
        }

        if (processedImages.isNotEmpty) {
          _identityMap[personName] = processedImages;
        }
      }
      if (scannedFiles > numOfFiles) break;
      scannedFiles++;
    }
    print("Loaded ${_identityMap.length} identities.");
  }

  // 🔥 FIX 3: Change return type to Float32List for high-performance memory packing
  Float32List _processImage(File file) {
    final bytes = file.readAsBytesSync();
    final raw = img.decodeImage(bytes);
    if (raw == null) throw Exception("Could not decode ${file.path}");

    final resized = img.copyResize(raw, width: imageSize, height: imageSize);

    // We use a flat Float32List to avoid the overhead of nested lists/Iterables
    final floatData = Float32List(imageSize * imageSize * 3);
    int idx = 0;

    for (var pixel in resized) {
      floatData[idx++] = pixel.r / 255.0;
      floatData[idx++] = pixel.g / 255.0;
      floatData[idx++] = pixel.b / 255.0;
    }
    return floatData;
  }

  Map<String, Float32List> nextBatch(int batchSize) {
    final int features = imageSize * imageSize * 3;
    final anchorBatch = Float32List(batchSize * features);
    final positiveBatch = Float32List(batchSize * features);
    final negativeBatch = Float32List(batchSize * features);

    final people = _identityMap.keys.toList();

    for (int i = 0; i < batchSize; i++) {
      // 1. Pick Anchor/Positive
      final personA = people[_random.nextInt(people.length)];
      final imagesA = _identityMap[personA]!;

      final idx1 = _random.nextInt(imagesA.length);
      int idx2 = _random.nextInt(imagesA.length);
      if (imagesA.length > 1) {
        while (idx1 == idx2) {
          idx2 = _random.nextInt(imagesA.length);
        }
      }

      // 2. Pick Negative
      String personB;
      do {
        personB = people[_random.nextInt(people.length)];
      } while (personA == personB);
      final imagesB = _identityMap[personB]!;
      final idx3 = _random.nextInt(imagesB.length);

      // 3. Set pixel data into the batch lists (Types now match: Float32List to Float32List)
      anchorBatch.setAll(i * features, imagesA[idx1]);
      positiveBatch.setAll(i * features, imagesA[idx2]);
      negativeBatch.setAll(i * features, imagesB[idx3]);
    }

    return {
      'anchor': anchorBatch,
      'positive': positiveBatch,
      'negative': negativeBatch,
    };
  }
}

File: python/dart_cuda/loaders/triplet_loader2.py
"""Triplet loader (RAM-cached float arrays) — mirrors
`lib/loaders/triplet_loader2.dart`.

Requires Pillow.
"""

from __future__ import annotations

import array
import os
import random
from typing import Dict, List


def _require_pil():
    try:
        from PIL import Image  # noqa: F401

        return Image
    except ImportError as exc:  # pragma: no cover
        raise ImportError("Pillow is required: pip install Pillow") from exc


class TripletLoader:
    def __init__(self, root_path: str, image_size: int, num_of_files: int):
        self.root_path = root_path
        self.image_size = image_size
        self._identity_map: Dict[str, List["array.array"]] = {}
        self._random = random.Random()
        self._scan_dataset(num_of_files)

    def _process_image(self, file_path: str) -> "array.array":
        Image = _require_pil()
        img = (
            Image.open(file_path)
            .convert("RGB")
            .resize((self.image_size, self.image_size))
        )
        buf = array.array("f")
        for r, g, b in img.getdata():
            buf.append(r / 255.0)
            buf.append(g / 255.0)
            buf.append(b / 255.0)
        return buf

    def _scan_dataset(self, num_of_files: int) -> None:
        if not os.path.isdir(self.root_path):
            return
        print("Caching images to RAM as float arrays...")
        scanned = 0
        for name in os.listdir(self.root_path):
            full = os.path.join(self.root_path, name)
            if os.path.isdir(full):
                processed: List["array.array"] = []
                for fname in os.listdir(full):
                    if fname.lower().endswith((".jpg", ".jpeg", ".png")):
                        processed.append(
                            self._process_image(os.path.join(full, fname))
                        )
                if processed:
                    self._identity_map[name] = processed
            if scanned > num_of_files:
                break
            scanned += 1
        print(f"Loaded {len(self._identity_map)} identities.")

    def next_batch(self, batch_size: int) -> Dict[str, "array.array"]:
        features = self.image_size * self.image_size * 3
        anchor = array.array("f", [0.0] * (batch_size * features))
        positive = array.array("f", [0.0] * (batch_size * features))
        negative = array.array("f", [0.0] * (batch_size * features))
        people = list(self._identity_map.keys())
        for i in range(batch_size):
            person_a = people[self._random.randrange(len(people))]
            images_a = self._identity_map[person_a]
            idx1 = self._random.randrange(len(images_a))
            idx2 = self._random.randrange(len(images_a))
            if len(images_a) > 1:
                while idx1 == idx2:
                    idx2 = self._random.randrange(len(images_a))
            person_b = person_a
            while person_b == person_a:
                person_b = people[self._random.randrange(len(people))]
            images_b = self._identity_map[person_b]
            idx3 = self._random.randrange(len(images_b))

            anchor[i * features : (i + 1) * features] = images_a[idx1]
            positive[i * features : (i + 1) * features] = images_a[idx2]
            negative[i * features : (i + 1) * features] = images_b[idx3]
        return {"anchor": anchor, "positive": positive, "negative": negative}

    nextBatch = next_batch
